{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from pyspark import SparkContext, SparkConf\n",
    "from pyspark.sql import SQLContext\n",
    "from pyspark.sql import HiveContext\n",
    "import json\n",
    "import time\n",
    "import sys\n",
    "from  scipy.stats import fisher_exact, ttest_ind\n",
    "import csv\n",
    "\n",
    "content = [line.rstrip() for line in open('jobsArguments.conf')]\n",
    "\n",
    "analysisName=content[0]\n",
    "scope=content[1]\n",
    "scale=content[2]\n",
    "sqlControl=content[3]\n",
    "sqlCase=content[4]\n",
    "controlGroupName=content[5]\n",
    "caseGroupName=content[6]\n",
    "controlMAF=content[7]\n",
    "caseMAF=content[8]\n",
    "pathVariants=content[9]\n",
    "scoringFunction=content[10]\n",
    "\n",
    "nPartitions=8\n",
    "conf = (SparkConf()\n",
    "         .setMaster(\"local[\"+str(nPartitions)+\"]\")\n",
    "       )\n",
    "#sc.stop()\n",
    "sc = SparkContext(conf=conf)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#sqlContext = HiveContext(sc) #sqlContext._get_hive_ctx() #HiveContext(sc) \n",
    "sqlContext = SQLContext(sc)\n",
    "sqlContext.sql(\"SET spark.sql.parquet.binaryAsString=true\")\n",
    "\n",
    "#pathVariants='/user/hive/warehouse/digest.db/exomes_hc_ulb'\n",
    "parquetFile = sqlContext.read.parquet(pathVariants)\n",
    "parquetFile.registerTempTable(\"variantData\");\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Input is vector patient, chr, pos, ref, alt, gene_symbol, zygosity\n",
    "def createKey_VariantGene(variantData):\n",
    "    #ID is chr:pos:ref:alt\n",
    "    ID=variantData[1]+\":\"+str(variantData[2])+\":\"+variantData[3]+\":\"+variantData[4]\n",
    "    \n",
    "    #return ID, gene_symbol, patient, zygosity\n",
    "    zygosity=1\n",
    "    if variantData[6]==\"Homozygous\":\n",
    "    #if variantData[6]==2:\n",
    "        zygosity=2\n",
    "    patientsID_dictionnary=patientsID_dictionnary_b.value\n",
    "    patientIndex=patientsID_dictionnary[variantData[0]]\n",
    "    return ((ID,variantData[5]),(patientIndex,zygosity))\n",
    "\n",
    "#variantGeneEntry: key is (variantID,gene), value is (patientIndex,zygosity)\n",
    "def geneAsKey(variantGeneEntry):    \n",
    "    return (variantGeneEntry[0][1],(variantGeneEntry[0][0],variantGeneEntry[1]))\n",
    "\n",
    "def getVariantID(key_VariantGene):\n",
    "    return key_VariantGene[0]\n",
    "\n",
    "def f(splitIndex ,v): \n",
    "    return [(splitIndex,list(v))]\n",
    "\n",
    "def toSQLString(strList):\n",
    "    strList=\"','\".join(strList)\n",
    "    return \"('\"+strList+\"')\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Transform sparse data (list of (sample_id,zygozity)) into vector z_i\n",
    "def vectorize(genotypeDataList):\n",
    "    genotypeDataList=list(genotypeDataList)\n",
    "    genotypeVector=[0]*len(patientsID_dictionnary_b.value)\n",
    "    if len(genotypeDataList)>0:\n",
    "        for j in range(0,len(genotypeDataList)):\n",
    "            genotypeVector[genotypeDataList[j][0]]=genotypeDataList[j][1]\n",
    "        \n",
    "        sumCase=float(sum([int(x>0) for x in genotypeVector[0:patientsID_split_index_b.value]]))\n",
    "        sumControl=float(sum([int(x>0) for x in genotypeVector[patientsID_split_index_b.value:len(patientsID_dictionnary_b.value)]]))\n",
    "    \n",
    "        ratioCase=sumCase/patientsID_split_index_b.value\n",
    "        ratioControl=sumControl/(len(patientsID_dictionnary_b.value)-patientsID_split_index_b.value)\n",
    "        \n",
    "        if (ratioCase>float(caseMAF_b.value)) or (ratioControl>float(controlMAF_b.value)):\n",
    "            genotypeVector=[0]*len(patientsID_dictionnary_b.value)\n",
    "        \n",
    "    return genotypeVector        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Compute burden for variantList\n",
    "def burden(geneID_variantList):\n",
    "    (geneID,variantList)=geneID_variantList\n",
    "    variantList=list(variantList)\n",
    "    burden=[0]*len(patientsID_dictionnary_b.value)\n",
    "    \n",
    "    if len(variantList)>0:\n",
    "        #Go through list of variants\n",
    "        for i in range(0,len(variantList)):\n",
    "            #Get variant ID, and list of sample_index,genotype\n",
    "            (variantID,genotypeDataList)=variantList[i]\n",
    "            #Get genotype vector for current variantID\n",
    "            genotypeDataVector=vectorize(genotypeDataList)\n",
    "            #And sum with previous genotype vectors\n",
    "            burden=[x+y for x,y in zip(burden,genotypeDataVector)]\n",
    "    \n",
    "    return (geneID,burden)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#variantList is [(locusID,[genotype])]\n",
    "def scoreVariant(ID_genotypeDataList):\n",
    "    ((variantID,geneID),genotypeDataList)=ID_genotypeDataList\n",
    "    #genotypeList=list(value_GenotypeList)\n",
    "    \n",
    "    patientsID_dictionnary=patientsID_dictionnary_b.value\n",
    "    patientsID_split_index=patientsID_split_index_b.value\n",
    "    \n",
    "    genotypeDataVector=vectorize(genotypeDataList)\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumpatients\":\n",
    "        sumCase=float(sum([int(x>0) for x in genotypeDataVector[0:patientsID_split_index]]))\n",
    "        sumControl=float(sum([int(x>0) for x in genotypeDataVector[patientsID_split_index:len(patientsID_dictionnary)]]))\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumalleles\":\n",
    "        sumCase=float(sum([x for x in genotypeDataVector[0:patientsID_split_index]]))\n",
    "        sumControl=float(sum([x for x in genotypeDataVector[patientsID_split_index:len(patientsID_dictionnary)]]))\n",
    "    \n",
    "    ratioCase=sumCase/patientsID_split_index\n",
    "    ratioControl=sumControl/(len(patientsID_dictionnary)-patientsID_split_index)\n",
    "        \n",
    "    score=ratioCase-ratioControl\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumpatients\":\n",
    "        pvalue=fisher_exact([[sumCase,patientsID_split_index-sumCase],[sumControl,len(patientsID_dictionnary)-patientsID_split_index]],'greater')[1]\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumalleles\":\n",
    "        pvalue=ttest_ind(genotypeDataVector[0:patientsID_split_index],genotypeDataVector[patientsID_split_index:len(patientsID_dictionnary)])[1]/2\n",
    "    \n",
    "    \n",
    "    if score>0:\n",
    "        return (variantID,geneID,score,pvalue,ratioCase,ratioControl,sumCase,sumControl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def scoreVariantPair(block_i,block_k,i,k):\n",
    "    block_k=list(block_k)\n",
    "    len_i=len(block_i)\n",
    "    len_k=len(block_k)\n",
    "    scores=[]\n",
    "\n",
    "    patientsID_dictionnary=patientsID_dictionnary_b.value\n",
    "    patientsID_split_index=patientsID_split_index_b.value\n",
    "    \n",
    "    start_k=0\n",
    "    skip_last=0\n",
    "    if i==k:\n",
    "        skip_last=1\n",
    "        if len_i==1:\n",
    "            len_i=0\n",
    "    \n",
    "    if len_i>0 and len_k>0:\n",
    "        for it_i in range(0,len_i-skip_last):\n",
    "            if i==k:\n",
    "                start_k=it_i+1\n",
    "            for it_k in range(start_k,len_k):\n",
    "                listLoadBlock_i=block_i[it_i]\n",
    "                listLoadBlock_k=block_k[it_k]\n",
    "                \n",
    "                genoSum=[int(x>0 and y>0) for x,y in zip(vectorize(listLoadBlock_i[1]),vectorize(listLoadBlock_k[1]))]\n",
    "                sumCase=float(sum([int(x>0) for x in genoSum[0:patientsID_split_index]]))\n",
    "                sumControl=float(sum([int(x>0) for x in genoSum[(patientsID_split_index):len(patientsID_dictionnary)]]))\n",
    "        \n",
    "                ratioCase=sumCase/patientsID_split_index\n",
    "                ratioControl=sumControl/(len(patientsID_dictionnary)-patientsID_split_index)\n",
    "        \n",
    "                score=ratioCase-ratioControl\n",
    "                pvalue=fisher_exact([[sumCase,patientsID_split_index-sumCase],[sumControl,len(patientsID_dictionnary)-patientsID_split_index]],'greater')[1]\n",
    "        \n",
    "                if score>=0:\n",
    "                    scores.append((listLoadBlock_i[0][0],listLoadBlock_i[0][1],listLoadBlock_k[0][0],listLoadBlock_k[0][1],score,pvalue,ratioCase,ratioControl,sumCase,sumControl))\n",
    "    return scores\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#variantList is [(locusID,[sample_index,genotype])]\n",
    "def scoreGene(geneID_burden):\n",
    "    (geneID,burden)=geneID_burden\n",
    "    \n",
    "    patientsID_dictionnary=patientsID_dictionnary_b.value\n",
    "    patientsID_split_index=patientsID_split_index_b.value\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumpatients\":\n",
    "        sumCase=float(sum([int(x>0) for x in burden[0:patientsID_split_index]]))\n",
    "        sumControl=float(sum([int(x>0) for x in burden[patientsID_split_index:len(patientsID_dictionnary)]]))\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumalleles\":\n",
    "        sumCase=float(sum([int(x) for x in burden[0:patientsID_split_index]]))\n",
    "        sumControl=float(sum([int(x) for x in burden[patientsID_split_index:len(patientsID_dictionnary)]]))\n",
    "    \n",
    "    ratioCase=sumCase/patientsID_split_index\n",
    "    ratioControl=sumControl/(len(patientsID_dictionnary)-patientsID_split_index)\n",
    "        \n",
    "    score=ratioCase-ratioControl\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumpatients\":\n",
    "        pvalue=fisher_exact([[sumCase,patientsID_split_index-sumCase],[sumControl,len(patientsID_dictionnary)-patientsID_split_index]],'greater')[1]\n",
    "    \n",
    "    if scoringFunction_b.value==\"sumalleles\":\n",
    "        pvalue=ttest_ind(burden[0:patientsID_split_index],burden[patientsID_split_index:len(patientsID_dictionnary)])[1]/2\n",
    "        \n",
    "    if score>=0:\n",
    "        return (geneID,score,pvalue,ratioCase,ratioControl,sumCase,sumControl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def scoreGenePair(block_i,block_k,i,k):\n",
    "    block_k=list(block_k)\n",
    "    len_i=len(block_i)\n",
    "    len_k=len(block_k)\n",
    "    scores=[]\n",
    "\n",
    "    patientsID_dictionnary=patientsID_dictionnary_b.value\n",
    "    patientsID_split_index=patientsID_split_index_b.value\n",
    "    \n",
    "    start_k=0\n",
    "    skip_last=0\n",
    "    if i==k:\n",
    "        skip_last=1\n",
    "        if len_i==1:\n",
    "            len_i=0\n",
    "    \n",
    "    if len_i>0 and len_k>0:\n",
    "        for it_i in range(0,len_i-skip_last):\n",
    "            if i==k:\n",
    "                start_k=it_i+1\n",
    "            for it_k in range(start_k,len_k):\n",
    "                listLoadBlock_i=block_i[it_i]\n",
    "                listLoadBlock_k=block_k[it_k]\n",
    "                genoSum=[int(x>0 and y>0) for x,y in zip(listLoadBlock_i[1],listLoadBlock_k[1])]\n",
    "                sumCase=float(sum([int(x>0) for x in genoSum[0:patientsID_split_index]]))\n",
    "                sumControl=float(sum([int(x>0) for x in genoSum[(patientsID_split_index):len(patientsID_dictionnary)]]))\n",
    "                #genoSum=[int(x+y) for x,y in zip(listLoadBlock_i[1],listLoadBlock_k[1])]\n",
    "                #sumCase=float(sum([int(x) for x in genoSum[0:patientsID_split_index]]))\n",
    "                #sumControl=float(sum([int(x) for x in genoSum[(patientsID_split_index):len(patientsID_dictionnary)]]))\n",
    "        \n",
    "                ratioCase=sumCase/patientsID_split_index\n",
    "                ratioControl=sumControl/(len(patientsID_dictionnary)-patientsID_split_index)\n",
    "        \n",
    "                score=ratioCase-ratioControl\n",
    "                pvalue=fisher_exact([[sumCase,patientsID_split_index-sumCase],[sumControl,len(patientsID_dictionnary)-patientsID_split_index]],'greater')[1]\n",
    "                #pvalue=ttest_ind(genoSum[0:patientsID_split_index],genoSum[patientsID_split_index:len(patientsID_dictionnary)])[1]/2\n",
    "    \n",
    "                if score>=0:\n",
    "                    scores.append((listLoadBlock_i[0],listLoadBlock_k[0],score,pvalue,ratioCase,ratioControl,sumCase,sumControl))\n",
    "    return scores\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ranking(sqlCase,sqlControl,scale,scope,p):\n",
    "    start_time = time.time()\n",
    "    nvariants=0\n",
    "    \n",
    "    variants_case = sqlContext.sql(\"SELECT sample_id,chr,pos,ref,alt,gene_symbol,zygosity FROM variantData \"+sqlCase)\n",
    "    variants_control= sqlContext.sql(\"SELECT sample_id,chr,pos,ref,alt,gene_symbol,zygosity FROM variantData \"+sqlControl)\n",
    "\n",
    "    variants=variants_control.unionAll(variants_case)\n",
    "    variants_grouped=variants.map(createKey_VariantGene).groupByKey(p)\n",
    "\n",
    "    if scope=='monogenic':\n",
    "        if scale=='variant':\n",
    "            ntests=variants_grouped.count()\n",
    "            nvariants=ntests\n",
    "            finish_load_time=time.time()\n",
    "            runtime_load=finish_load_time - start_time\n",
    "            scores=variants_grouped.map(scoreVariant).filter(lambda x:x is not None).takeOrdered(1000, key=lambda (variant,gene,v1,v2,v3,v4,v5,v6): -v1)\n",
    "\n",
    "        if scale=='gene':\n",
    "            variants_grouped_by_gene=variants_grouped.map(geneAsKey).groupByKey(p)\n",
    "            ntests=variants_grouped_by_gene.count()\n",
    "            finish_load_time=time.time()\n",
    "            runtime_load=finish_load_time - start_time\n",
    "            burden_by_gene=variants_grouped_by_gene.map(burden)\n",
    "            burden_by_gene.count()\n",
    "            finish_burden_time=time.time()\n",
    "            runtime_burden=finish_burden_time - finish_load_time\n",
    "            scores=burden_by_gene.map(scoreGene).filter(lambda x:x is not None).takeOrdered(1000, key=lambda (k,v1,v2,v3,v4,v5,v6): -v1)\n",
    "\n",
    "    if scope=='digenic':\n",
    "        if scale=='variant':\n",
    "            ntests=variants_grouped.count()\n",
    "            nvariants=ntests\n",
    "            finish_load_time=time.time()\n",
    "            runtime_load=finish_load_time - start_time\n",
    "            \n",
    "            variants_grouped_with_partitions=variants_grouped.mapPartitionsWithIndex(lambda splitIndex,v: [(splitIndex,list(v))])\n",
    "            scores=[]\n",
    "            for i in range(0,p):\n",
    "                block_i=variants_grouped_with_partitions.filter(lambda (k,v):k==i).collect()[0][1]\n",
    "                score=variants_grouped_with_partitions.filter(lambda (k,v):k>=i).flatMap(lambda (k,v):scoreVariantPair(block_i,v,i,k)).takeOrdered(1000, key=lambda (variant1,gene1,variant2,gene2,v1,v2,v3,v4,v5,v6): -v1)\n",
    "                scores=scores+score\n",
    "            scores=sc.parallelize(scores,p).takeOrdered(1000, key=lambda (variant1,gene1,variant2,gene2,v1,v2,v3,v4,v5,v6): -v1)\n",
    "            ntests=ntests*(ntests+1)/2\n",
    "   \n",
    "        if scale=='gene':\n",
    "            variants_grouped_by_gene=variants_grouped.map(geneAsKey).groupByKey(p)\n",
    "            ntests=variants_grouped_by_gene.count()\n",
    "            finish_load_time=time.time()\n",
    "            runtime_load=finish_load_time - start_time\n",
    "            burden_by_gene=variants_grouped_by_gene.map(burden)\n",
    "            burden_by_gene.count()\n",
    "            finish_burden_time=time.time()\n",
    "            runtime_burden=finish_burden_time - finish_load_time\n",
    "            burden_by_gene_with_partitions=burden_by_gene.mapPartitionsWithIndex(lambda splitIndex,v: [(splitIndex,list(v))])\n",
    "            #burden_by_gene_with_partitions.cache()\n",
    "            scores=[]\n",
    "            for i in range(0,p):\n",
    "                block_i=burden_by_gene_with_partitions.filter(lambda (k,v):k==i).collect()[0][1]\n",
    "                score=burden_by_gene_with_partitions.filter(lambda (k,v):k>=i).flatMap(lambda (k,v):scoreGenePair(block_i,v,i,k)).takeOrdered(1000, key=lambda (gene1,gene2,v1,v2,v3,v4,v5,v6): -v1)\n",
    "                scores=scores+score\n",
    "            scores=sc.parallelize(scores,p).takeOrdered(1000, key=lambda (gene1,gene2,v1,v2,v3,v4,v5,v6): -v1)\n",
    "            ntests=ntests*(ntests+1)/2\n",
    "    \n",
    "    end_time=time.time()\n",
    "    if scale=='variant':\n",
    "        runtime_score=end_time - finish_load_time\n",
    "        all_times=[runtime_load,0,runtime_score]\n",
    "    if scale==\"gene\":\n",
    "        runtime_score=end_time - finish_burden_time\n",
    "        all_times=[runtime_load,runtime_burden,runtime_score]\n",
    "\n",
    "    return (all_times,scores,ntests,nvariants)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p=10\n",
    "runtimes=[]\n",
    "start_time=time.time()\n",
    "\n",
    "patientsID_case = sqlContext.sql(\"SELECT distinct sample_id FROM variantData \"+sqlCase).collect()\n",
    "patientsID_case = [patients[0] for patients in patientsID_case]\n",
    "\n",
    "patientsID_control = sqlContext.sql(\"SELECT distinct sample_id FROM variantData \"+sqlControl).collect()\n",
    "patientsID_control = [patients[0] for patients in patientsID_control]\n",
    "\n",
    "patientsID=patientsID_case+patientsID_control\n",
    "patientsID_dictionnary=dict(zip(patientsID,range(len(patientsID))))\n",
    "\n",
    "patientsID_split_index_b = sc.broadcast(len(patientsID_case))\n",
    "patientsID_dictionnary_b = sc.broadcast(patientsID_dictionnary)\n",
    "                     \n",
    "controlMAF_b=sc.broadcast(controlMAF)\n",
    "caseMAF_b=sc.broadcast(caseMAF)\n",
    "\n",
    "scoringFunction_b=sc.broadcast(scoringFunction)\n",
    "\n",
    "end_time=time.time()\n",
    "(all_times,scores,ntests,nvariants)=ranking(sqlCase,sqlControl,scale,scope,p)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "metadata=[analysisName,scale,scope,sqlControl,sqlCase,patientsID_control,patientsID_case,controlGroupName,caseGroupName,ntests,start_time,end_time,all_times]\n",
    "\n",
    "with open(analysisName+'_metadata.json', 'w') as outfile:\n",
    "    json.dump(metadata, outfile)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open(analysisName+'_ranking.csv', 'w') as csvfile:\n",
    "    csvwriter = csv.writer(csvfile, delimiter=' ',\n",
    "                            quotechar='\"', quoting=csv.QUOTE_MINIMAL)\n",
    "    csvwriter.writerows(scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sc.stop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
